### PR TITLE
[FIX] http_routing: commit assets also when generating an error page

### PR DESCRIPTION
**Pull request by bso-odoo**
Since [1] (HTTPocalypse) the generated asset bundles are rolled back
whenever an error page is generated.
This causes the error page to have a wrong layout if it is rendered
before any non-error page.

This commit informs generated `assetsbundle` that they should be saved
when an error HTML response it generated.

Steps to reproduce:
- Start an instance with "-i website".
- Access "/doesnotexist" before accessing any other URL.

=> Error page layout was broken because it could not obtain its assets.

[1]: https://github.com/odoo/odoo/commit/723f1f0d119285b6409032719240d4658633b616

task-2990132

### CODE REVIEW

**Comment by bso-odoo** [comment]
@Julien00859 Hi, I made the assets saved upon error page generation after previous operations were rolled back. I believe this is the right location for this but I'd rather have your opinion too.

File: addons/http_routing/models/ir_http.py [hunk 1]
```diff
@@ -637,6 +637,12 @@ def _handle_error(cls, exception):
         code, values = cls._get_exception_code_values(exception)
 
         request.cr.rollback()
+        # After changes were rolled back, if the error generates assets bundle
+        # attachments, those must be committed for the error to be properly
+        # displayed.
+        context = dict(request.env.context)
+        context['commit_assetsbundle'] = True
+        request.env = request.env(request.env.cr, request.env.uid, context)
```
Comment (Julien00859) [inline addons/http_routing/models/ir_http.py:645]: Please don't reassign the `request.env` attribute yourself, use `request.update_env` or `request.update_context` instead
```suggestion
        request.update_context(commit_assetsbundle=True)
```

**Review by Julien00859 — COMMENTED** [review]
Frankly all the bits around this context key are still obscure to this day to me and we plan to refactor the wa assetbundles are generated for readonly soon, what do you think @Xavier-Do ?

**Comment by Xavier-Do** [comment]
> Frankly all the bits around this context key are still obscure to this day to me and we plan to refactor the wa assetbundles are generated for readonly soon, what do you think @Xavier-Do ?

Yes I totally agree, I should make this a priority actually. It was almost ready but got distracted. 

But this is an issue and the won't be solved in stable so this may remain a problem for 16.0

Anyway it should only happen if the 404 is the first request. 

There is another possible fix, it is the `add_post_rollback` I introduced in https://github.com/odoo/odoo/pull/121376
If the transaction is rollback, it clears the ormcache. Backporting this should partially solve this issue in 16.0.

**Comment by Julien00859** [comment]
Can you test using the post-rollback instead of the commit? https://github.com/odoo-dev/odoo/blob/ab86b63b27ddc691b10376b75cced2a46745b90f/odoo/addons/base/models/assetsbundle.py#L305

**Comment by Julien00859** [comment]
So using post-rollback, *until you render any other page* the 404 ones will be broken. But if you load a valid page after a 404 then it'll be correctly rendered.

At the moment if the first ever loaded page is the 404 one, then it has a broken layout and *any future valid page* have a broken layout too.

So this wouldn't fix the broken layout on the firsts 404 but we rather have that that introduce yet another extra commit (which play very badly with readonly)

**Comment by bso-odoo** [comment]
@Julien00859 I was looking into it... but as far as I can tell it solves a different issue:
With the current one-liner fix, we make sure the assets are created, therefore not needing the post-rollback hook to clear the cache.
With the post-rollback hook we clean the cache so that upon further requests the assets get generated.

In our case, I believe we want the the 404 page to be displayed correctly - which requires its assets. Therefore I think the `commit_assetsbundle=True` is the way to go.

**Comment by Xavier-Do** [comment]
commit_assetsbundle is a hack that was introduced for reports and we are trying to get rid of it. This is why wae are trying to resist adding one more. 

The layout broken on the first 404 is sad, but not critical. The really problematic part IMHO is to have a the layout broken on all other following frontend call. 

If it is required to have the layout fixed for the 404 called before any valid frontend page, we should just generate the page in a committed transaction, not using an obscure context hack  :) But this is such a small corner case that I would say that we don't care.

**Comment by rdeodoo** [comment]
> The layout broken on the first 404 is sad, but not critical. The really problematic part IMHO is to have a the layout broken on all other following frontend call.

As I mentioned on the task spec, this would be acceptable :+1: 
As long as the client business is not completely killed if the first page is a 404 and visitors can still navigate to legit pages, it's okay.

**Comment by bso-odoo** [comment]
FYI - I am still investigating why the test manages to pass without the fix.
I get different results when I test manually.
I neutralized any `--dev=xml` just in case.
I digged into the `t-cache` mechanism, and I could not find a difference there.
I could not find my intruder by tracking `@ormcache` either.

**Comment by Julien00859** [comment]
> FYI - I am still investigating why the test manages to pass without the fix.

Surely because there are shenanigans in tests about the transactions

**Comment by Julien00859** [comment]
> I am still investigating why the test manages to pass without the fix.
The test is working fine on my machine using the latest commit, it fails without the fix and works with the fix, are you still experimenting the issue?

**Comment by bso-odoo** [comment]
> The test is working fine on my machine using the latest commit, it fails without the fix and works with the fix, are you still experimenting the issue?

I just tried again and it now properly fails without the fix... no idea what I was doing wrongly yesterday.
Sorry for the induced workload and delay.

**Comment by Julien00859** [comment]
no problem, can you fix that ci/style so we can merge?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Just passing by before someone merge this

File: odoo/addons/base/models/assetsbundle.py [hunk 2]
```diff
@@ -321,6 +321,14 @@ def get_attachments(self, extension, ignore_version=False):
         attachment_ids = [r[0] for r in self.env.cr.fetchall()]
         return self.env['ir.attachment'].sudo().browse(attachment_ids)
 
+    def add_post_rollback(self):
+        """
+            In some rare cases it is possible that an attachment is created during a transaction, added to the ormcache
+            but the transaction is rollbacked, leading to 404 when getting the attachments.
+            This postrollback hook will help fix this issue by clearing the cache if it isn't commited.
+        """
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/models/assetsbundle.py:329]: While we are at it, pep8 advocates for a 72 char limit for "long chunk of continued text", i.e. docstrings, logging and exceptions mostly. Also, it is more common to do docstrings this way, notice no extra indentation:

```py
def foo():
    """
    Long docstring
    """
```

**Review by Julien00859 — APPROVED** [review]

**Review by rdeodoo — APPROVED** [review]
@robodoo r+
@fw-bot up to saas-16.3 (it's fixed in 16.4 where the backport of this pr comes from)

I just changed the commit message, you mentioned https://github.com/odoo/odoo/commit/188cf158fc72368165257a4cef1b62187d66d881 instead of https://github.com/odoo/odoo/commit/6d5d234f15e458c4a92c240fc743aee941b460e2 which is the real commit (main issue with the odoo-dev one is that it does not have the link to the PR which is what we usually need)

**Pull request closed without merging**